packages/kbn-utility-types/index.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { PromiseType } from 'utility-types';
+
+/**
+ * Returns wrapped type of a promise.
+ */
+export type UnwrapPromise<T extends Promise<any>> = PromiseType<T>;
+
+/**
+ * Minimal interface for an object resembling an `Observable`.
+ */
+export interface ObservableLike<T> {
+  subscribe(observer: (value: T) => void): void;
+}
+
+/**
+ * Returns wrapped type of an observable.
+ */
+export type UnwrapObservable<T extends ObservableLike<any>> = T extends ObservableLike<infer U>
+  ? U
+  : never;
+
+/**
+ * Same as `Promise` type, but it flat maps the wrapped type.
+ */
+export type ShallowPromise<T> = T extends Promise<infer U> ? Promise<U> : Promise<T>;
```

Comment:
I'm trying to come up with a situation where I'd use this type, but I can't come up with anything. I'm thinking maybe something like `Promise.resolve()`, but I'm struggling... Do you have a usage in mind?

Reply:
When an async function returns something, and you return it again

![image](https://user-images.githubusercontent.com/9773803/61373787-03e13980-a89b-11e9-9467-e859fcd4fa95.png)